### PR TITLE
ldap search with paging

### LINKED ISSUE #7702: LDAP search failed: Result Code 4 "Size Limit Exceeded"
**Describe the bug**
LDAP queries that return a large number of results are still erroring with the fix for #4162 . When an LDAP server is setup to only return 1000 results a group query which returns more than that will fail.

**To Reproduce**
1. Setup LDAP auth with a group filter that will return more than 1000 groups
2. Attempt to login
3. Login fails with `LDAP search failed: LDAP Result Code 4 "Size Limit Exceeded"`

**Expected behavior**
The login should be successful and all groups should be returned in the query response so authentication will succeed.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.2.3
* Vault CLI Version (retrieve with `vault version`): 1.2.3
* Server Operating System/Architecture: Linux

Vault server configuration file(s):

```
vault write auth/ldap/config \
  url="ldaps://ldap.example.com:636" \
  userdn="OU=Users,OU=Org,OU=BusinessUnits,DC=example,DC=com" \
  groupdn="OU=AccountGroups,OU=Groups,OU=Org,OU=BusinessUnits,DC=example,DC=com" \
  groupfilter="(&(objectClass=group))" \
  groupattr="memberOf" \
  upndomain="example.com" \
  insecure_tls=false \
  starttls=false

vault write auth/ldap/groups/developers policies=dev
```

**Additional context**
To see the problem, the LDAP server must be setup to return fewer groups than the number of groups which match the query.

This search appears to need updated: https://github.com/hashicorp/vault/blob/master/sdk/helper/ldaputil/client.go#L120-L125

Should use something like this:

```
result, err := conn.SearchWithPaging(&ldap.SearchRequest{
  BaseDN:    cfg.UserDN,
  Scope:     ldap.ScopeWholeSubtree,
  Filter:    filter,
  SizeLimit: math.MaxInt32,
}, math.MaxInt32)
```

Example using go-ldap. Given the proper group query, the first search will fail and the second will succeed:

```
package main

import (
	"fmt"
	"log"
	"math"

	"github.com/go-ldap/ldap"
)

func main() {
	l, err := ldap.Dial("tcp", "ldap.example.com:389")
	if err != nil {
		log.Fatal(err)
	}
	defer l.Close()
	l.Bind("<binduser>", "<bindpw>")
	request := &ldap.SearchRequest{
		BaseDN: "OU=AccountGroups,OU=Groups,OU=Org,OU=BusinessUnits,DC=example,DC=com",
		Scope:  ldap.ScopeWholeSubtree,
		Filter: "(&(objectClass=group))",
		Attributes: []string{
			"member:1.2.840.113556.1.4.1941:",
		},
		SizeLimit: math.MaxInt32,
	}

	//ERROR: SIZE LIMIT EXCEEDED
	sr, err := l.Search(request)
	if err != nil {
		log.Print(err)
	} else {
		for _, entry := range sr.Entries {
			fmt.Printf("%s: %v\n", entry.DN, entry.GetAttributeValue("cn"))
		}
	}

	//WORKS
	sr, err = l.SearchWithPaging(request, math.MaxInt32)
	if err != nil {
		log.Fatal(err)
	}

	for _, entry := range sr.Entries {
		fmt.Printf("%s: %v\n", entry.DN, entry.GetAttributeValue("cn"))
	}
}
```

### PR DESCRIPTION
**Pull request by jackielii**
this fixes #7702

### CODE REVIEW

**Comment by jackielii** [comment]
already done the cla, somehow it still shows not signed

**Comment by tyrannosaurus-becks** [comment]
Hi @jackielii , thank you for providing this pull request! 

To err on the side of caution, would it be possible to add an LDAP configuration parameter of `ldap_search_page_size` and default it to 0? Then the logic would be, if it's set to > 0, use pagination, otherwise don't? 

The reason is, I'd like to maintain the exact current behavior for folks unless they _elect_ to set the page size. Then we can fix the problem, while also being 100% sure we won't break anything for anyone.

**Comment by tyrannosaurus-becks** [comment]
Also, I'm not sure why the CLA signing isn't working, could there be a mismatch between the email that Github thinks was used for this pull request, and the email that Github thinks you're using to sign it? Or maybe it was a transient failure? Would you be willing to try again?

Sorry for the trouble with it! :-)

**Comment by jackielii** [comment]
@tyrannosaurus-becks no probs. Just amended the commit message for the author email.

when it comes to `ldap_search_page_size`, as the comment says:

```
// LDAPSearchPageSize is the page size to search LDAP directory.
// Default size is usually 1000, but this number doesn't affect the result.
// It only affect the number of round trips the client has to do.
var LDAPSearchPageSize uint32 = 500
```

The paging is done within the call `conn.SearchWithPaging`, it only affects how many round trips it has to use, it always returns the full set as a slice. I know, this feels like a wired API.

**Comment by ncabatoff** [comment]
Hi @jackielii,

I agree with @tyrannosaurus-becks, I think for backwards-compatibility reasons we should make the new behaviour opt-in with a config option.  There are so many different LDAP/AD implementations out there that I'm afraid without the opt-in behaviour that we're likely to have a regression somewhere.

**Comment by jackielii** [comment]
> Hi @jackielii,
> 
> I agree with @tyrannosaurus-becks, I think for backwards-compatibility reasons we should make the new behaviour opt-in with a config option. There are so many different LDAP/AD implementations out there that I'm afraid without the opt-in behaviour that we're likely to have a regression somewhere.

Sorry for the late reply, as I mentioned in the earlier comments: this doesn't change the behavior of the search, it only allows search against a bigger directory.

I think the confusion comes from the SearchWithPaging vs Search: if you look at the API for [SearchWithPaging](https://pkg.go.dev/github.com/go-ldap/ldap/v3?tab=doc#Conn.Search). It states that:

> SearchWithPaging accepts a search request and desired page size in order to execute LDAP queries to fulfill the search request. All paged LDAP query responses will be buffered and the final result will be returned atomically. 

> A requested pagingSize of 0 is interpreted as no limit by LDAP servers.

`Search` is basically SearchWithPaging with pagingSize of 0. This might be fine for some LDAP implementations, but on the MS Active Directory it triggers an error `LDAP Result Code 4 "Size Limit Exceeded"`

So in summary, `SearchWithPaging` is really a safer `Search`

**Comment by useful-devops-tools** [comment]
@ncabatoff  and @jackielii 

Is there any update on this Pull request?
We're running into the same issue now:
`LDAP Result Code 4 "Size Limit Exceeded`

It looks like this PR would solve the issue with pagination.

**Comment by ncabatoff** [comment]
> So in summary, SearchWithPaging is really just a safer Search

I am anything but an expert in LDAP, so take this with a grain of salt, but: I don't believe the above is true.  At a high level, perhaps, but in practice from looking at rfc2696 and the openldap client code linked, SearchWithPaging uses an LDAP extension and does not issue the same request as Search does just with a 0 instead of a non-zero page size.  The RFC is over 20y old now, probably everyone supports this extension at this point, but my experience with LDAP implementations is that they have subtle differences that can come back to bite you.   

I maintain my belief that we should make this opt-in behaviour using a new config option.  In any event, isn't it desirable to be able to configure the paging size?

**Comment by ltcarbonell** [comment]
Closing out this PR, work duplicated by the change [here](https://github.com/hashicorp/vault/pull/17640)
